### ISSUE TITLE
Empty XML with results from DelegatingXmlCreationSink

### ISSUE THREAD

**Issue by sobkowicz** [xunit/xunit#2090]
I have problem with generated XML after running tests. I expect valid XML with tests results which I get from `DelegatingXmlCreationSink` but instead sometimes this XML is empty and results appear after a short time (I think 10-20 ms).

I use custom AssemblyRunner that use `IFrontController`, `DelegatingExecutionSummarySink` and `DelegatingXmlCreationSink` to run tests and get generated XML with results.
Details you can find in sample project (only key features to reproduce the problem).

**Sample app:**
https://github.com/sobkowicz/xunit-sink-xml-app

**Steps to reproduce:**
Download and run sample app without debuging. Empty XML appears only sometimes so I added runnning tests in loop, you can increase loop count. I added saving empty XMLs to stream and then to a file. About 1000 loops is enough to get empty XML in most cases.

**Expected XML:**
```
<assembly name="ExternalTests.dll"
          environment="64-bit .NET Standard [collection-per-class, parallel (8 threads)]"
          test-framework="xUnit.net 2.4.1.0"
          run-date="2020-04-14"
          run-time="12:03:00"
          total="1"
          passed="1"
          failed="0"
          skipped="0"
          time="0.001"
          errors="0">
    <errors />
    <collection total="1"
                passed="1"
                failed="0"
                skipped="0"
                name="Test collection for ExternalTests.ExampleTests"
                time="0.000">
        <test name="ExternalTests.ExampleTests.EmptyAndPassingTest"
              type="ExternalTests.ExampleTests"
              method="EmptyAndPassingTest"
              time="0.0000036"
              result="Pass" />
    </collection>
</assembly>
```

**XML in some cases:**
```
<assembly name="ExternalTests.dll"
          environment="64-bit .NET Standard [collection-per-class, parallel (8 threads)]"
          test-framework="xUnit.net 2.4.1.0"
          run-date="2020-04-14"
          run-time="12:03:00">
    <errors />
</assembly>
```

xunit version: 2.4.1
xunit.runner.utility: 2.4.1
xunit.extensibility.core: 2.4.1
.NET Core version: 2.1
OS: Windows 10

**Comment by bradwilson** [comment]
Oof, given how simple the repro is, this feels like it might be a race condition in the reporting. :frowning:

**Comment by kdthomas2121** [comment]
Is there any update to this? Experiencing the same 25% of the time

**Comment by bjorn-stromberg-codic** [comment]
This still happens regularly in our CI as well two years later. Is there any good work around?

**Comment by bradwilson** [comment]
No, the bug is still present as far as I know.

**Comment by jonlegendre** [comment]
Hi @bradwilson - 
I believe I have identified a root cause and fix for this issue. I am happy to create a PR (or PRs), however, I am uncertain which branch(es) you would like PRs against for the fix. Can you let me know? I see the issue on both main (v3) and older branches. 

Happy to chat in a PR, but I do have a concern about the potential fix:

From what I can see, the issue is caused in the class `DelegatingXmlCreationSink`, in the method `OnMessage`. Since the `innerSink.OnMessage(message)` is being invoked prior to `HandleTestCollectionFinished` being invoked, sinks (including the one passed into the `XunitFrontController` constructor) are being passed the the `_TestCollectionFinished`(v3) \ `ITestCollectionFinished`(older versions) message potentially prior to the XML being added in the `HandleTestCollectionFinished` method. These two have a race condition depending on how fast the xml is created in `HandleTestCollectionFinished` versus how quickly the message is consumed by the sink(s). 

However, I am concerned, because you have a relevant comment in the code:
```
		// Call the inner sink first, because we want to be able to depend on ExecutionSummary
		// being correctly filled out.
		var result = innerSink.OnMessage(message);
```
So this seems intentional, however, the fix is to move the `innerSink.OnMessage(message)` into the place where the result is instead i.e.:
```
		return message.DispatchWhen<_ErrorMessage>(HandleErrorMessage)
			//several lines skipped for brevity
			&& message.DispatchWhen<_TestCollectionFinished>(HandleTestCollectionFinished)
			//several lines skipped for brevity
			&& message.DispatchWhen<_TestStarting>(HandleTestStarting)
			//&& result; -- COMMENTED OUT FOR CLARITY
			&& innerSink.OnMessage(message);
```
If this fix won't work due to the commented concern, another solution could be to add an additional message something like `_TestCollectionResultsAvailable` and allow folks to subscribe to that instead to avoid the overloaded `_TestCollectionFinished`; however, I am not sure if that is too large of a change to consider for this issue.
Alternatively, I could create logic to have special behaviors for the `_TestCollectionFinished` message, but I suspect that message is the very one that concerns your comment.

Happy to proceed with any pattern you suggest.

**Comment by bradwilson** [comment]
@jonlegendre writes:

> Hi @bradwilson - I believe I have identified a root cause and fix for this issue. I am happy to create a PR (or PRs), however, I am uncertain which branch(es) you would like PRs against for the fix. Can you let me know? I see the issue on both main (v3) and older branches.

We would definitely need code for both branches (whether you fixed one and I ported it to the other, or you fixed both, is up to you).

> From what I can see, the issue is caused in the class `DelegatingXmlCreationSink`, in the method `OnMessage`. Since the `innerSink.OnMessage(message)` is being invoked prior to `HandleTestCollectionFinished` being invoked, sinks (including the one passed into the `XunitFrontController` constructor) are being passed the the `_TestCollectionFinished`(v3) \ `ITestCollectionFinished`(older versions) message potentially prior to the XML being added in the `HandleTestCollectionFinished` method. These two have a race condition depending on how fast the xml is created in `HandleTestCollectionFinished` versus how quickly the message is consumed by the sink(s).
> 
> However, I am concerned, because you have a relevant comment in the code:
> 
> ```
> 		// Call the inner sink first, because we want to be able to depend on ExecutionSummary
> 		// being correctly filled out.
> 		var result = innerSink.OnMessage(message);
> ```
> 
> So this seems intentional

The comment is related to knowledge that `DelegatingExecutionSummarySink` always lives at the "top" of the stack and is needed to ensure that the execution summary is completely filled out in `HandleTestAssemblyFinished`. We could move the inner sink invocation down as you suggest, with the caveat that we call `HandleTestAssemblyFinished` after the inner sink has had a chance to run.

It's very likely given the way things are currently structured that there are potential XML-dependency related race conditions for any message, and we just happen to be seeing one with TestCollectionFinished.

The more I think about it, the more I believe that the comment in `DelegatingXmlCreationSink` is an indication that I put things in the stack in the wrong manner. Instead of:

`DelegatingExecutionSummarySink`
  `DelegatingXmlCreationSink`
    `DelegatingLongRunningTestDetectionSink`
      `DelegatingFailSkipSink`

It should probably be:

`DelegatingXmlCreationSink`
  `DelegatingExecutionSummarySink`
    `DelegatingLongRunningTestDetectionSink`
      `DelegatingFailSkipSink`

And then ensure we move calling the inner sink to after we've done our own processing in all cases. That would allow `DelegatingXmlCreationSink` to just assume that things have been filled out correctly (which they would be, since `DelegatingExecutionSummarySink` would run first).

This is not as simple as it sounds, since these delegators implement an additional interface (`IExecutionSink`), and they all assume they delegate to another `IExecutionSink` _except_ for `DelegatingExecutionSummarySink`, because it assumes it will always be at the top. Creation order has always been assumed to be fixed in this way, which is why the comment in `DelegatingXmlCreationSink` exists.

So trying to make changes here would at the very least necessitate creating breaking changes in the constructors of these delegating sinks, which are publicly shipped classes. We could introduce an entirely new set of classes which do the same things, but with different names, so that we can obsolete the old ones but don't remove them, because they may be in use in third party runners.

(As an aside, I have a local branch on my box where I've been exploring the idea of getting rid of `IExecutionSink` entirely in v3. This might be the time to do so here, because I found the idea that `IExecutionSink` is implemented in most cases to just be "I don't know, ask the other one" to be unnecessary. If we go through with a plan like "let's make all new classes", it would be an ideal time to get rid of usage of `IExecutionSink` as well and just let them all be delegators for `IMessageSinkWithTypes` instead.)

Let me ponder on whether there is a less intrusive way to fix this problem.

The alternative that's close to what you suggest would be to move the innerSink invocation near the bottom of the list, except for the invocation of dispatching TestAssemblyFinished, which is what relies on the innerSink being invoked first. Additionally, `DelegatingExecutionSummarySink` right now invokes the innerSink at the top of its stack as well, which is probably wrong for the same reasons (but less impactfully).

> 		return message.DispatchWhen<_ErrorMessage>(HandleErrorMessage)
> 			//several lines skipped for brevity
> 			&& message.DispatchWhen<_TestCollectionFinished>(HandleTestCollectionFinished)
> 			//several lines skipped for brevity
> 			&& message.DispatchWhen<_TestStarting>(HandleTestStarting)
> 			//&& result; -- COMMENTED OUT FOR CLARITY
> 			&& innerSink.OnMessage(message);

Something that's important to remember here is that we can't just put the call to the inner sink in the `&&` chain here because the inner sink must _always_ run, even when a message above has called for cancellation. I have verified in the v2 source in all four delegating sinks in xunit.runner.utility that they are all careful to ensure the inner sink is always called (and I assume that continues into the v3 source, as I don't remember making many changes here other than updating for the new types).

> Note: This may also be related to #1534, although I have not investigated to be sure.

Agreed, this seems likely related.

**Comment by jonlegendre** [comment]
Thanks Brad.
I will hold on any further action for now. Note that I did test the fix I suggested and it resolved the issue for the example provided by @sobkowicz, but as I suspected due to the comment we discussed above, the correct solution is a bit more complicated.
Please let me know if you would like me to take action on a less intrusive approach.

**Comment by bradwilson** [comment]
@jonlegendre Feel free to open a PR so I can see. Thanks!

**Comment by bradwilson** [comment]
I decided that the best strategy would be to just stop the delegating chain pattern and combine everything into a single ExecutionSink: https://github.com/xunit/xunit/commit/b41e8ad2b9010e36edbc87bc8fbf8cc634c2d682

The fact that dependency order was an issue made all the delegation designs I prototyped unattractive. In truth the core functionality (which came from `DelegatingExecutionSummarySink`) is really quite simple: count errors, and record the assembly results. Everything else (fail skips, long running test detection, XML generation) is optional, and only is run when the user has requested it. It's probably _slighty_ slower as a result of the conditionals, though unlikely to be noticed in the noise of how long tests take to run.

I verified this with the OP's test project (after converting it to `net6.0`). I would get between 1-3 failures previously running 1,000 iterations (and I printed out the loop #; it _always_ failed on the first iteration of the loop, plus 0-2 additional failures). After applying this fix, I was unable to get it to fail even with 10,000 iterations. I tried 100,000 but I think due to the fact that everything is being stashed in memory I was hitting excessive swapping. 😂 

This is available in v2 core framework version `2.6.3-pre.3`.

It's also integrated into `xunit.runner.visualstudio` version `2.5.5-pre.3` though that's mostly irrelevant since it's incapable of generating native XML output: https://github.com/xunit/visualstudio.xunit/commit/56db5de6cece6ca1341664ceece0a1e186dceca8
